### PR TITLE
Fix the conflict with tap gesture in UploadKeys page

### PR DESCRIPTION
**Pull request by LucasXu0**
Hope this helps

### CODE REVIEW

File: ExposureNotification/ViewControllers/UploadKeysViewController.swift [hunk 1]
```diff
@@ -321,7 +311,9 @@ class UploadKeysViewController: UIViewController {
             $0.edges.equalToSuperview()
         }
         
-        view.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard)))
+        let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard))
+        tapGestureRecognizer.delegate = self
```
Comment (ShivaHuang) [inline ExposureNotification/ViewControllers/UploadKeysViewController.swift]: Do you mind declare `tapGestureRecognizer` in the form like `stackView` at line 254 and move to line around 308, just after `stackView`?

So we can clearly identify how the recognizer is configured and make all components declared at the beginning of function.

**Comment by PichuChen** [comment]
Will you have testing clip for this fix?

**Comment by LucasXu0** [comment]
Yes. Here is the demo video.

https://user-images.githubusercontent.com/11863087/125883700-64818e24-8287-4de8-b699-ddcadee1bc7b.mp4

File: ExposureNotification/ViewControllers/UploadKeysViewController.swift [hunk 2]
```diff
@@ -415,6 +407,25 @@ extension UploadKeysViewController: UITextFieldDelegate {
...
+        let location = gestureRecognizer.location(in: gestureRecognizer.view)
+        
+        for view in whiteListViews {
+            let point = view.convert(location, from: gestureRecognizer.view)
+            if view.point(inside: point, with: nil) {
+                return false
+            }
+        }
+        
+        return true
```
Comment (ShivaHuang) [inline ExposureNotification/ViewControllers/UploadKeysViewController.swift]: I think we don't need the `location`. How about converting location directly onto `whiteListViews`? Something like

```swift
for view in whiteListViews {
    let point = gestureRecognizer.location(in: view)
    ...
}
```

And think functional, we can do like following

```swift
        return whiteListViews.map {
            let location = gestureRecognizer.location(in: $0)
            // Break the chain if touching inside views of white list.
            return $0.point(inside: location, with: nil) == false
        }.reduce(true) { $0 && $1 }
```

File: ExposureNotification/ViewControllers/UploadKeysViewController.swift [hunk 3]
```diff
@@ -415,6 +407,25 @@ extension UploadKeysViewController: UITextFieldDelegate {
     }
 }
 
+extension UploadKeysViewController: UIGestureRecognizerDelegate {
+    func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
+        
+        // The views that should not be responsed to the tap gesture.
```
Comment (ShivaHuang) [inline ExposureNotification/ViewControllers/UploadKeysViewController.swift]: There is a violation of grammar.

This sentence should be
```
// The views that should not response to the tap gesture.
```

**Review by ShivaHuang — CHANGES_REQUESTED** [review]

File: ExposureNotification/ViewControllers/UploadKeysViewController.swift (hunk 1 above)
Comment (LucasXu0) [reply ExposureNotification/ViewControllers/UploadKeysViewController.swift]: Sure. I have submitted the new commit.

File: ExposureNotification/ViewControllers/UploadKeysViewController.swift (hunk 3 above)
Comment (LucasXu0) [reply ExposureNotification/ViewControllers/UploadKeysViewController.swift]: Thanks.

File: ExposureNotification/ViewControllers/UploadKeysViewController.swift (hunk 2 above)
Comment (LucasXu0) [reply ExposureNotification/ViewControllers/UploadKeysViewController.swift]: Yes, don't need the `convert` function. And I have summited the new code (more functional).

Comment (LucasXu0) [reply ExposureNotification/ViewControllers/UploadKeysViewController.swift]: Actually, for the functions that need to return directly, I think the original code may be better, reducing the times of map the array.

Comment (ShivaHuang) [reply ExposureNotification/ViewControllers/UploadKeysViewController.swift]: You use `allSatisfy` which is much better than I suggested. Thanks!

And about the performance, There are only two pickers, so I'm not worried about that.

**Comment by ShivaHuang** [comment]
All looks good for me.

However, I'm not member of AI Labs anymore, so I don't have the permission to merge this PR anymore.

Please be patient, I think someone will handle it later.

**Comment by nicesoap99** [comment]
There's a layout issue for date picker in smaller device like iPhone 8.
I think create an `UIScrollView` as the superview of `stackView` is a way to solve it.
Would you mind to fix it?

Thank you.

**Comment by LucasXu0** [comment]
When debugging, I found that Xcode will output the following error.

```
2021-07-20 12:53:49.204940+0800 ExposureNotification[11557:212143] [DatePicker] UIDatePicker 0x7fd7531391c0 is being laid out below its minimum width of 280. This may not look like expected, especially with larger than normal font sizes.
```

I have no idea how to fix this layout error currently. It seems to be an `inline` style problem. try to create an `UIScrollView ` to contain the `stackView`, but there will be another layout problem.

Please let me know if you have any ideas to fix the layout error.

**Comment by nicesoap99** [comment]
After some tests, I think the error is related to the `isHidden` property of `UIDatePicker`.
I don't know why but there is a workaround.

You can re-arrange the subView of `stackView` when `editingField` `didSet` instead of making it hidden

For example, after tapping the `startDateButton`, you may do something like below:
```
    stackView.insertArrangedSubview(startDatePicker, belowArrangedSubview: startDateButton)

    stackView.removeArrangedSubview(endDatePicker)
    endDatePicker.removeFromSuperview()
```
Here is a extension method for more convenient
```
extension UIStackView {
    func insertArrangedSubview(_ view: UIView, belowArrangedSubview subview: UIView) {
        arrangedSubviews.enumerated().forEach {
            if $0.element == subview {
                insertArrangedSubview(view, at: $0.offset + 1)
            }
        }
    }
}
```

Thank you.

**Comment by LucasXu0** [comment]
I have tested the change that you mentioned, but still can't fix the problem.

After some investigation, I think this error occurs because of nested `UIStackView`. `UIDatePicker` with `inline` style is also a `UIStackView`, and something goes wrong when Apple handles these scenes.

截屏2021-07-22 下午2 18 09

截屏2021-07-22 上午11 07 00


Maybe keep the original `wheels` style is fine. 😂

**Comment by nicesoap99** [comment]
Honestly, I think the `wheels` style is fine too. 😉

But if you are interested in fixing this, please check this [stackOverflow](https://stackoverflow.com/questions/33073127/nested-uistackviews-broken-constraints)

Thank you.

**Comment by LucasXu0** [comment]
I may have found the cause of the problem.

```
stackView.snp.makeConstraints {
    $0.top.equalTo(view.safeAreaLayoutGuide).offset(60)
    $0.left.right.equalTo(view.safeAreaLayoutGuide)
    $0.bottom.equalTo(view.safeAreaLayoutGuide).offset(-79)
}
```

截屏2021-07-22 下午6 01 04
There is no enough space for inner `UIDatePicker` to display.

After change the external `UIStackView` layout, the `UIDatePicker` could display good. But this will broke the another `View` layout.
```
stackView.snp.makeConstraints {
    $0.top.equalTo(0)
    $0.left.right.equalTo(view.safeAreaLayoutGuide)
    $0.bottom.equalTo(view.safeAreaLayoutGuide)
}
```

截屏2021-07-22 下午6 05 05

**Comment by LucasXu0** [comment]
In fact, the `UIDatePicker` with other style (like `wheel`) may also cause this problem, but their height is relatively low. So there is no impact for the time being.

**Pull request closed without merging**